autoload/clap.vim

```diff
@@ -182,12 +182,13 @@ endfunction
 
 " Sometimes we don't need to go back to the start window, hence clap#_exit() is extracted.
 function! clap#exit() abort
-  call clap#_exit()
+  noautocmd call clap#_exit()
 
   " NOTE: Need to go back to the start window
   if win_getid() != g:clap.start.winid
     call g:clap.start.goto_win()
   endif
+  doautocmd WinEnter
```

Comment:
I think this patch should be enough?

```diff
diff --git a/autoload/clap.vim b/autoload/clap.vim
index 70b0897..b4ae9ac 100644
--- a/autoload/clap.vim
+++ b/autoload/clap.vim
@@ -187,6 +187,7 @@ function! clap#exit() abort
   " NOTE: Need to go back to the start window
   if win_getid() != g:clap.start.winid
     call g:clap.start.goto_win()
+    doautocmd WinEnter
   endif
 endfunction
```

Reply:
It does work yes, on the other hand if we let autocommands being triggered during `clap#_exit()` I feel that we're adding unnecessary work. Because there will be autocommands triggered for staying in a window that will not keep the focus. Those autocommands perform work, such as updating tabline and statusline plugin. It's small delays, but it adds to the perceived perfomance of clap because it increases the time before opening the new buffer.

I also have another proposition:
```vim
" Sometimes we don't need to go back to the start window, hence clap#_exit() is extracted.
function! clap#exit() abort
  noautocmd call g:clap.start.goto_win() " noautocmd is repeated in goto_win() but is here for readability
  noautocmd call clap#_exit()
  doautocmd WinEnter
endfunction
```

I think this might be even better because it is functionally equivalent, and it completely avoids focusing a window that isn't needed, we go directly back to the previous window, then cleanup, then trigger the autocommand.

Reply:
>Those autocommands perform work, such as updating tabline and statusline plugin.

That's my convern. I'm not 100% sure whethe it's appropriate to disable these autocmds, some plugins might be broken like your barbar.nvim, so I previously keep using `noautocmd` as much as less even I do care about the performance. I'm not saying I'm against it, it's definitely good for the performance. Let's try it for a while and see if something expected happens.

Reply:
And what was the rationale for marking `clap.start.goto_win()` as `noautocmd`, if you remember? And do you have some example use-cases of plugins where this would cause perturbations? I could try a few of them see that we're not causing unexpected behavior. IMHO the current implementation is good, I've been studying autocommands in the last few days due to this plugin and [that PR](https://github.com/neovim/neovim/pull/13117) and I'm confident in the current behavior.

One thing I've realized just now is that there are other exit points from clap such as `clap#selection#try_open`. Let's leave this open I'll test them in the next days.

Reply:
>And what was the rationale for marking clap.start.goto_win() as noautocmd, if you remember? 

Hmm, if I recall correctly, using `noautocmd` inside `g:clap.start.goto_win()` is simply to avoid writing `noautocmd` when calling `g:clap.start.goto_win()` :). So adding `doautocmd WinEnter` is indeed what I missed before.

>And do you have some example use-cases of plugins where this would cause perturbations?

https://github.com/liuchengxu/vim-clap/blob/9a89a257a713b93e309c1ab82d38085f0185bed9/autoload/clap.vim#L158

If someone has some `WinClosed` autocmd config in his vimrc, then I believe it won't be triggered.